Add documentation for DATEADD function in OQL (#4)
dahfjkg/docs: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md

²In OQL v1, the expression gets the type of the first argument. If you use OQL v1, the type of `AgeOrAmount` in the example above is Integer, not Decimal.

### DATEADD {#dateadd-function}

The `DATEADD` function adds a specified period of time to an expression of type `DATETIME`. The return type is `DATETIME`.

> This function was introduced in Mendix version 11.9.0

Reviewer: I think we should add a disclaimer for only runtime support like we did for STRING_AGG:

`This function is only supported in Java actions.`

#### Syntax

The syntax is as follows:

```sql
DATEADD ( datepart , length_expression , date_expression [, timezone ] )
```

##### datepart

> `datepart` specifies the interval in which `length_expression` is measured. For possible values, see [Examples](#oql-dateadd-example), below.

Reviewer: we are missing YEAR, SECOND and WEEK in the example. Could we also a reference [DATEPART](https://docs.mendix.com/refguide/oql-expression-syntax/#oql-datepart-example) for the full list?

| 2025-03-02 15:12:45 | 2024-05-29 15:12:45 | 2024-06-02 15:27:45 |

The optional time zone parameter affects the difference between standard time and daylight saving time. For example, let's assume that a user in the time zone `Europe/Berlin` is running the OQL query below. In 2024, daylight saving time in that time zone ended on 27th of October at 02:00. Before that moment, the offset was `UTC+2h`, and after that it was `UTC+1h`. In `UTC` time zone, there was no clock change, and the offset was always `UTC+0h`. That explains the 1-hour difference between the results of the `DATEADD` function with the default time zone value `Europe/Berlin` and with the explicitly specified value `UTC`.

```sql
SELECT
	Start,
	DATEADD(MONTH, 6, Start) AS StartPlus6MonthsBerlin,
	DATEADD(MONTH, 6, Start, 'UTC') AS StartPlus6MonthsUTC,
FROM
	Sales.Period
WHERE
	Revenue = 28
```

| Start               | StartPlus6MonthsBerlin  | StartPlus6MonthsUTC  |                                                        
|:--------------------|-------------------------|---------------------:|
| 2024-05-02 00:00:00 | 2024-11-02 00:00:00     | 2024-11-01 23:00:00  |

> {{% alert color="info" %}}

Reviewer: Perhaps its better to place this near the datapart syntax section?

²In OQL v1, the expression gets the type of the first argument. If you use OQL v1, the type of `AgeOrAmount` in the example above is Integer, not Decimal.

### DATEADD {#dateadd-function}

The `DATEADD` function adds a specified period of time to an expression of type `DATETIME`. The return type is `DATETIME`.

> This function was introduced in Mendix version 11.9.0. It is currently supported only in Java actions.

Reviewer 2: ```suggestion
{{% alert color="info" %}}
This function was introduced in Mendix version 11.9.0. It is supported only in Java actions.
{{% /alert %}}
```

#### Syntax

The syntax is as follows:

```sql
DATEADD ( datepart , length_expression , date_expression [, timezone ] )
```

##### datepart

`datepart` specifies the interval in which `length_expression` is measured.

Supported options are `YEAR`, `QUARTER`, `MONTH`, `WEEK`, `DAY`, `HOUR`, `MINUTE`, `SECOND`.

##### length_expression

> `length_expression` specifies the amount of intervals to add or subtract. For example, `DATEADD ( WEEK , 3 , OrderDate )` adds 3 weeks to an attribute `OrderDate`. Expressions that resolve to Integer or Long are allowed.

Reviewer 2: For precision perhaps
```suggestion
`length_expression` specifies the number of `datepart` intervals to add. For example, `DATEADD ( WEEK , 3 , OrderDate )` adds 3 weeks to an attribute `OrderDate`. You can provide a negative number to subtract a period. Expressions that resolve to Integer or Long are allowed.
```

##### date_expression

`date_expression` specifies the date to add intervals to. The expression should resolve to a `DATETIME` value. String representations of `DATETIME` are accepted.

##### timezone

`timezone` specifies the time zone to use for the operation. This parameter is optional and defaults to the user time zone. It should be a string literal containing an IANA time zone. GMT offset time zones are not supported.

> For the `DATEADD` function, this parameter affects the difference between standard time and daylight saving time.

Reviewer 2: This felt unclear until I realized there is an example below which makes things clearer.
```suggestion
For the `DATEADD` function, this parameter affects the difference between standard time and daylight saving time. See [Examples](#oql-dateadd-example), below.
```

The user time zone is usually different from UTC. To get the result in the UTC time zone, explicitly specify `'UTC'` in this parameter. For details on time zone handling in Mendix Runtime, see [Date and Time Handling](/refguide/date-and-time-handling/).
{{% /alert %}}

#### Examples{#oql-dateadd-example}

Assume the entity `Sales.Period` has 2 objects:

```sql
SELECT * FROM Sales.Period
```

| ID | Start               | End                 | Revenue |                                                        
|:---|---------------------|---------------------|--------:|
| -  | 2024-05-02 00:00:00 | 2025-07-05 00:00:00 | 28      |
| -  | 2024-05-02 00:00:00 | 2024-06-02 15:12:45 | 10      |

You can use `DATEADD` to add or subtract intervals from date values:

```sql
SELECT
	DATEADD(QUARTER, 3, End) AS EndPlus3Quarters,
	DATEADD(DAY, -5, End) AS EndMinus5Days,
	DATEADD(MINUTE, 15, End) AS EndPlus15Minutes,
FROM
	Sales.Period
```

| EndPlus3Quarters    | EndMinus5Days       | EndPlus15Minutes    |                                                        
|:--------------------|---------------------|--------------------:|
| 2026-04-05 00:00:00 | 2025-06-30 00:00:00 | 2025-07-05 00:15:00 |

> | 2025-03-02 15:12:45 | 2024-05-29 15:12:45 | 2024-06-02 15:27:45 |

Reviewer 2: Is this EndPlus3Quarters a good example? It looks to me like if the local timezone has summer time then the answer could be 2025-03-02 14:12:45 as it is going from DST to non-DST?

Author: The calculation is performed in the user time zone, and it is displayed to the user in the same time zone. Indeed, `2025-03-02 15:12:45` has 1 hour difference with the original date in UTC, but the time should be the same in user time zone. Thus, the example is correct, but, as it often happens with time zones, not clear. The difference is explained in the second example where `DATEADD(MONTH, 6, Start, 'UTC')` returns a different result. If that is not clear enough, let's add a better explanation

Reviewer 2: Thanks. So long as the example is correct that's fine. The customer should put in some edge cases in their testing to make sure it's giving the result they expect 🙂

| "Performance is satisfactory" | 27          |
| "Order has been completed"    | 24          |

### LOCATE{#locate-function}

#### Description

> Returns the index of the first occurence of a substring in a string. If no occurrence is found, returns 0.

Reviewer: The examples and default value make this clear, but lets still explicitly mention the index is 1-based

> Returns the index of the first occurence of a substring in a string. The index is 1-based. If no occurrence is found, returns 0.

Reviewer 2: ```suggestion
Returns the index of the first occurrence of a substring in a string. The index is 1-based. If the substring is not in the string, it returns 0.
```

This function was introduced in Mendix version 11.9.0. It is currently supported only in Java actions.

#### Syntax

The syntax is as follows:

```sql
LOCATE ( expression , pattern [, offset ] )
```

> `expression` specifies the string to be searched.

Reviewer: Like in DATEADD let's be explicit on expressions of type string being allowed. Same for `pattern` and `offset` but with integer type.

| "Performance is satisfactory" | 27          |
| "Order has been completed"    | 24          |

### LOCATE{#locate-function}

#### Description

Returns the index of the first occurrence of a substring in a string. The index is 1-based. If the substring is not in the string, it returns 0.

This function was introduced in Mendix version 11.9.0. It is currently supported only in Java actions.

#### Syntax

The syntax is as follows:

```sql
LOCATE ( expression , pattern [, offset ] )
```

> `expression` specifies the string to be searched. Expressions that resolve to String are allowed.

Reviewer 2: Is there a reason why you call the second parameter `pattern` rather than `substring`.
Pattern implies to me that this might accept regular expressions - but the description says it is just a string. 
Would it be better to make it explicit that it is a string?

Author: I copied that from the `REPLACE` function. Changed both to substring

| 0.33         | 3.33333333 |
| 1.17         | 1.17142857 |

### SUBSTRING{#substring-function}

#### Description

Returns part of a string starting at specified character index.

This function was introduced in Mendix version 11.9.0. It is currently supported only in Java actions.

#### Syntax

The syntax is as follows:

```sql
SUBSTRING ( expression , start_char [, length ] )
```

`expression` specifies the string to be searched. Expressions that resolve to String are allowed.

`start_char` specifies the 1-based index of the first character of the resulting substring. Expressions that resolve to Integer or Long are allowed.

> `length` specifies the length limit of the resulting substring. If the `length` argument is not specified, all characters starting with `start_char` are returned. Expressions that resolve to Integer or Long are allowed.

Reviewer 2: Would it be clearer to say "maximum length"

```suggestion
`length` specifies the maximum length of the resulting substring. If the result would be longer than `length` it will be truncated. If the `length` argument is not specified, all characters starting with `start_char` are returned. Expressions that resolve to Integer or Long are allowed.
```

#### Example

```sql
SELECT
	LastName,
	SUBSTRING(UPPER(LastName), 3, 2) AS Substring_3_2,
FROM Sales.Customer
```

| LastName | Substring_3_2 |
|:---------|:--------------|
| Doe      | E             |
| Moose    | OS            |

If the string is shorter than the value of `start_char`, the result is an empty string. If it is shorter than `start_char + length`, the result is shorter than `length`:

```sql
SELECT
	SUBSTRING('dendrochronological', 13) AS Substring_13,
	SUBSTRING('dendrochronological', 13, 5) AS Substring_13_5,
	SUBSTRING('dendrochronological', 13, 10) AS Substring_13_10,
	SUBSTRING('dendrochronological', 20) AS Substring_20,
FROM Sales.Customer
ORDER BY LastName LIMIT 1
```

| Substring_13 | Substring_13_5 | Substring_13_10 | Substring_20 |
|:-------------|:---------------|:----------------|:-------------|

> | `logical`    | `logic`        | `logical`       | ` `          |

Reviewer 2: Is Substring_20 correct? It looks like a single space, length 1. Shouldn't it be an empty string, length zero?

Author: In VS Code preview, ` ` looks like an empty string. But here on github it does not, so I'll come up with a better way to show that

| 0.33         | 3.33333333 |
| 1.17         | 1.17142857 |

### SUBSTRING{#substring-function}

#### Description

Returns part of a string starting at specified character index.

This function was introduced in Mendix version 11.9.0. It is currently supported only in Java actions.

#### Syntax

The syntax is as follows:

```sql
SUBSTRING ( expression , start_char [, length ] )
```

`expression` specifies the string to be searched. Expressions that resolve to String are allowed.

`start_char` specifies the 1-based index of the first character of the resulting substring. Expressions that resolve to Integer or Long are allowed.

> `length` specifies the length limit of the resulting substring. Expressions that resolve to Integer or Long are allowed.

Reviewer 3: I suggest mentioning explicitly that if the `length` argument is not specified, all of the input string from the `start_char` will be used.
